File: pleg.py
import numpy as np

#def PLEG(angleo,gmu,m,nexp,nzen,ypleg):

# g=0.85
# cosine of the solar zenith angle =0.9
# Gaussion quadrature (single); order 8+8
#
# S =            16
#  G =    0.8400000000000000
#  MU0 =     1.000000000000000
#  OMEGA =     1.000000000000000
#  FTOP =     1.000000000000000
#  TAU =     1.000000000000000
#
# nzen=8
# nexp=8
# angle0 = .9
# gmu = np.ones(nzen)
# ypleg = np.zeros((nexp,nzen+2))
# something=8
# for m in range(something):
#     if m==0:
#         cm1=1.
#         dm1=1.
#     else:
#         cm2 = cm1 * ((2*m-1)/(2*m))**(.5)
#         dm2 = dm1 * ((2*m+1)/(2*m))**(.5)
#
#     for i in range(nzen+2):
#         if i == nzen+1:
#             mu = 0
#         elif i==nzen:
#             mu = angle0
#         else:
#             mu = gmu[i-1]
#
#         if m==0:
#             ypleg[0,i] = 1.
#             ypleg[1,i] = mu
#         else:
#             ypleg[m,i] = cm2 * (1-mu**2)**(m/2.)
#             ypleg[m+1,i] = dm2 * mu * (1-mu**2)**(m/2.)
#
#         for il in range(1,nexp-1):
#             l=il-1
#             if l < m:
#                 ypleg[il,i] = 0
#             else:
#                 lm2 = (l-m+2)*(l+m+2)
#                 lm1 = (l-m+1)*(l+m+1)
#                 ypleg[il+1,i] = (((2*l+3)*mu*ypleg[il,i]) - ((lm1**.5) * ypleg[il-1,i])) / (lm2**.5)
#     if m>0:
#         cm1=cm2
#         dm1=dm2
#
# print(ypleg)
#
#
# def P(g,mu):
#     return (1-g**2)/( (1+g**2-2*g*mu)**(3./2))
#  N=2*NZ
#       DO 12 I=1,NZ
#         Z=DCOS(3.141592654D0*(I-.25D0)/(N+.5D0))
#

File: pleg.f90
      SUBROUTINE PLEG(ANGLEO,GMU,M,NEXP,NZEN,YPLEG)
C----------------------------------------------------------------------C
C      Computation of renormalized associated Legendre polynomials     C  
C      of the form:                                                    C
C                                                                      C
C                m                     1/2 m                           C
C               Y (u) = [(l-m)!/(l+m)!]   P (u)                        C
C                l                         l                           C
C                                                                      C
C         where,                                                       C
C                 m                                                    C
C                P (u) = Associated Legendre polynomial of             C
C                 l      order l and degree m                          C
C                                                                      C
C                    u = Cosine of zenith angle                        C
C                                                                      C
C     Reference:                                                       C
C                                                                      C
C             Dave, J. V.,and Armstrong, B. H., 1970: Computations     C
C                  of High-order Associated Legendre Polynomials,      C
C                  JQSRT, 10, 557-562, 1970                            C
C----------------------------------------------------------------------C
C                 I N P U T    V A R I A B L E S :                     C
C----------------------------------------------------------------------C
C         ANGLEO   :    Cosine of solar zenith angle                   C
C         GMU      :    Gaussian quadrature points                     C
C         M        :    Index for degree of Legendre polynomial        C
C         NEXP     :    Tot. no. of polynomial expansion terms         C
C         NZEN     :    Tot. no. of quadrature points                  C
C----------------------------------------------------------------------C
C                O U T P U T    V A R I A B L E S :                    C
C----------------------------------------------------------------------C
C         YPLEG    :    Renormalized associated Legendre polynomials   C
C----------------------------------------------------------------------C
      REAL  GMU(NZEN),LM1,LM2,MU,YPLEG(NEXP,NZEN+2)
      SAVE  CM1, DM1
C
      IF (M .EQ. 0) THEN
        CM1 = 1.
        DM1 = 1.
      ELSE  !We're only dealing with m=0 here I think, so these are irrelevent for now...
        CM2 = -SQRT((2.*M-1.)/(2.*M))*CM1
        DM2 = -SQRT((2.*M+1.)/(2.*M))*DM1
      END IF
      DO 20,I=1,NZEN+2
        IF (I .LE. NZEN) MU = GMU(I)
        IF (I .EQ. NZEN+1) MU = ANGLEO
        IF (I .EQ. NZEN+2) MU = 0.
C----------------------------------------------------------------------C
C                  Compute initial values                              C
C----------------------------------------------------------------------C
        IF (M .EQ. 0) THEN
          YPLEG(1,I) = 1.
          YPLEG(2,I) = MU       
        ELSE  
          YPLEG(M+1,I) = CM2*(1-MU**2)**(M/2.)
          YPLEG(M+2,I) = DM2*MU*(1-MU**2)**(M/2.)
        END IF
        DO 10,IL=1,NEXP-2   !Fills array from IL+2 (so 3, since 1 and 2 are done), last IL=NEXP-2 so YPLEG(NEXP) so L0,N-1
          L = IL-1
          IF (IL .LT. M+1) THEN   !the integral here is L = M,N so if M!=0, start at M
            YPLEG(IL,I) = 0.
            GO TO 10
          ELSE
C----------------------------------------------------------------------C
C          Use varying-degree recurrence formula to compute            C
C          successive values                                           C
C----------------------------------------------------------------------C
            LM2 = (L-M+2)*(L+M+2)
            LM1 = (L+M+1)*(L-M+1)
            YPLEG(IL+2,I) = ((2*L+3)*MU*YPLEG(IL+1,I)-SQRT(LM1)*        
     $            YPLEG(IL,I))/SQRT(LM2)
            !Since M=0...
            !Lm2 = (L-2)*(L+2)
            !LM1 = (L+1)*(L-1)
            !YPLEG(IL+2,I) = 1/(sqrt(LM2) * ( (2*L+3) * MU * Y(IL+1,I) -sqrt(LM1)*Y(IL,I))
            !if L=IL-1... IL = L+1   also, 2*L + 3 = 2(L+1)+1 = 2*IL+1
            !YPLEG(IL+2,I) = 1/(sqrt(LM2) * ( (2*IL+1) * MU * Y(IL+1,I) -sqrt(LM1)*Y(IL,I))
            !L-M ! / L+m ! **.5
            !but if m is 0...  this term is 1??
            !when IL=1, filling YPLEG(IL+2), Y(3) ... L=0 so filling N=2

          END IF
  10    CONTINUE
  20  CONTINUE
      IF (M .GT. 0) THEN 
        CM1 = CM2
        DM1 = DM2
      END IF 
      RETURN        
      END

